From b1e364f5bc0e6ce8280b956b134aa76dd8cfb6d4 Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Date: Wed, 14 Oct 2015 17:07:09 +0300
Subject: [PATCH 0750/1429] dpaa2-eth: Do not allow ethtool settings change
 while the NI is up

Due to a MC limitation, link state changes while the DPNI is enabled
will fail. For now, we'll just prevent the call from going down to the MC
if we know it will fail.

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
[Original patch taken from QorIQ-SDK-V2.0-20160527-yocto]
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
index cbae068..4365de9 100644
--- a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
+++ b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
@@ -134,6 +134,15 @@ static int __cold ldpaa_set_settings(struct net_device *net_dev,
 
 	netdev_info(net_dev, "Setting link parameters...");
 
+	/* Due to a temporary firmware limitation, the DPNI must be down
+	 * in order to be able to change link settings. Taking steps to let
+	 * the user know that.
+	 */
+	if (netif_running(net_dev)) {
+		netdev_info(net_dev, "Sorry, interface must be brought down first.\n");
+		return -EACCES;
+	}
+
 	cfg.rate = ethtool_cmd_speed(cmd);
 	if (cmd->autoneg == AUTONEG_ENABLE)
 		cfg.options |= DPNI_LINK_OPT_AUTONEG;
-- 
1.7.5.4

